airbyte-ci/connectors/pipelines/pipelines/dagger/actions/gradle.py

```diff
@@ -0,0 +1,230 @@
+
+        return wrapper
+
+    def warm_up_dependency_cache(
+        self, gradle_dep_cache_path: str, gradle_home_path: str, build_cdk: bool, warm_dependency_cache_args: Tuple[str] = ("--dry-run")
+    ) -> Callable:
+        """Closure to seed the gradle home cache from the cache volume.
+        Args:
+            gradle_dep_cache_path (str): The path to the gradle dependency cache volume mount.
+            gradle_home_path (str): The path to the gradle home directory.
+            build_cdk (bool): Whether to build the CDK.
+            local_maven_repository_path (str): The path to the local maven repository.
+
+        Returns:
+            Callable: A function which takes a gradle container and returns a gradle container with the gradle cache seeded from the volume.
+        """
+
+        rsync_cache_volume_to_gradle_home = f"(rsync -a --stats --mkpath {gradle_dep_cache_path}/ {gradle_home_path} || true)"
+        rsync_gradle_home_to_cache_volume = f"(rsync -a --stats {self.GRADLE_HOME_PATH}/ {self.GRADLE_DEP_CACHE_PATH} || true)"
+        # Running a gradle task like "help" with these arguments will trigger updating all dependencies.
+        # When the cache is cold, this downloads many gigabytes of jars and poms from all over the internet.
+        resolve_deps_and_write_checksum = self._get_gradle_command("help", *warm_dependency_cache_args)
+        build_and_publish_local_cdk = self._get_gradle_command(":airbyte-cdk:java:airbyte-cdk:publishSnapshotIfNeeded")
+
+        gradle_commands = [resolve_deps_and_write_checksum]
+        if build_cdk:
+            gradle_commands.append(build_and_publish_local_cdk)
+
+        commands = [
+            rsync_cache_volume_to_gradle_home,
+            *gradle_commands,
+            rsync_gradle_home_to_cache_volume,
+        ]
+
+        def wrapper(gradle_container: dagger.Container) -> dagger.Container:
+            return gradle_container.with_exec(sh_dash_c(commands))
+
+        return wrapper
+
+    def _get_gradle_command(self, task: str, *args) -> str:
+        """Get the gradlew command to run a given task with the given arguments."""
+        return f"./gradlew {' '.join(self.DEFAULT_GRADLE_TASK_OPTIONS + args)} {task}"
+
+    async def __aenter__(self) -> GradleTaskExecutor:
+        """Enter the context manager and return a gradle task executor with a base gradle container with a seeded cache and S3 build cache set up.
+
+        Returns:
+            GradleTaskExecutor: A gradle task executor with a base gradle container.
+        """
+        warm_dependency_cache_args = ["--write-verification-metadata", "sha256", "--dry-run"] if not self.local_execution else ["--dry-run"]
+
+        self.gradle_container = await (
+            self.get_base_gradle_container()
+            .with_(self.set_s3_build_cache_credentials(self.s3_build_cache_access_key_id, self.s3_build_cache_secret_key))
+            .with_mounted_directory(self.workdir, self.gradle_project_dir)
```

Comment:
This is not going to do what we want. You want to maintain the old behaviour where the whole repo is mounted to download _all_ the dependencies. This way, in practice, that task is a no-op. Otherwise the cache volume gets polluted from one run to the next depending on which connector you build and you end up downloading stuff, which is _slow_.

Please, maintain the existing behaviour exactly. It took me several days to get right because it's so, so full of gotchas.